File: matiasmasso/Mat.NET/DTO/DTOKpi.cs
using System.Collections.Generic;
using System.Linq;

namespace DTO
{
    public class DTOKpi
    {
        public List<DTOYearMonth> YearMonths { get; set; }
        public string Caption { get; set; }
        public decimal YFactor { get; set; }
        public Ids Id { get; set; }

        public enum Formats
        {
            Eur,
            Decimal
        }

        public enum Ids
        {
            Activo_Corriente,
            Pasivo_Corriente,
            Comandes_Proveidors,
            Comandes_Clients,
            Efectes_Vençuts,
            Efectes_Impagats,
            Index_Impagats
        }



        public static DTOKpi Factory(Ids id)
        {
            DTOKpi retval = new DTOKpi();
            retval.Id = id;
            retval.YearMonths = new List<DTOYearMonth>();
            return retval;
        }

        public Formats format()
        {
            Formats retval = Formats.Eur;
            switch (Id)
            {
                case Ids.Index_Impagats:
                    {
                        retval = Formats.Decimal;
                        break;
                    }
            }
            return retval;
        }

        public decimal MaxValue()
        {
            var retval = 0;
            if (YearMonths.Count > 0)
                retval = (int)YearMonths.Max(x => x.Eur);
            return retval;
        }

        public DTOYearMonth YearMonthFrom()
        {
            return DTOYearMonth.Min(YearMonths);
        }

        public DTOYearMonth YearMonthTo()
        {
            return DTOYearMonth.Max(YearMonths);
        }

        public int MonthsCount()
        {
            return DTOYearMonth.MonthsDiff(YearMonthFrom(), YearMonthTo());
        }

        public DTOYearMonth AddYearMonth(int year, DTOYearMonth.Months month, decimal eur)
        {
            DTOYearMonth retval = new DTOYearMonth(year, month, eur);
            YearMonths.Add(retval);
            return retval;
        }
    }
}

File: matiasmasso/Mat.NET/DT2/DTOKpi.vb
Public Class DTOKpi
    Property YearMonths As List(Of DTOYearMonth)
    Property Caption As String
    Property YFactor As Decimal
    Property Id As Ids

    Public Enum Formats
        Eur
        [Decimal]
    End Enum

    Public Enum Ids
        Activo_Corriente
        Pasivo_Corriente
        Comandes_Proveidors
        Comandes_Clients
        Efectes_Vençuts
        Efectes_Impagats
        Index_Impagats
    End Enum



    Shared Function Factory(id As Ids) As DTOKpi
        Dim retval As New DTOKpi
        retval.Id = id
        retval.YearMonths = New List(Of DTOYearMonth)
        Return retval
    End Function

    Public Function format() As Formats
        Dim retval As Formats = Formats.Eur
        Select Case _Id
            Case Ids.Index_Impagats
                retval = Formats.Decimal
        End Select
        Return retval
    End Function

    Public Function MaxValue() As Decimal
            Dim retval = 0
            If _YearMonths.Count > 0 Then
                retval = _YearMonths.Max(Function(x) x.Eur)
            End If
            Return retval
        End Function

        Public Function YearMonthFrom() As DTOYearMonth
            Return DTOYearMonth.Min(_YearMonths)
        End Function

        Public Function YearMonthTo() As DTOYearMonth
            Return DTOYearMonth.Max(_YearMonths)
        End Function

    Public Function MonthsCount() As Integer
        Return DTOYearMonth.MonthsDiff(YearMonthFrom, YearMonthTo)
    End Function

    Public Function AddYearMonth(year As Integer, month As Integer, eur As Decimal) As DTOYearMonth
        Dim retval As New DTOYearMonth(year, month, eur)
        _YearMonths.Add(retval)
        Return retval
    End Function
End Class

